### PR TITLE
Add RefineTypes JIT pass for Tuple

### PR DESCRIPTION
**Pull request by henrytwo**
Consider the following JIT graph, where the type of `%a` and `%b` are out of sync with tuple `%c`.
Before:
```
graph(%a : Float(123), %b : Float(4, 5, 6)):
    c : (Tensor, Tensor) = prim::TupleConstruct(%a, %b)
    return (%c)
```
After:
```
graph(%a : Float(123), %b : Float(4, 5, 6)):
    c : (Float(123), Float(4, 5, 6)) = prim::TupleConstruct(%a, %b)
    return (%c)
```
This PR adds a pass `RefineTypes(...)` to update all such instances with the correct type. This is also available via Python by using `torch._C._jit_pass_refine_types(...)`.

A unit test has been added for unnamed tuples, but no test exists for `NamedTuple` (though it was tested manually) since it isn't supported by the parser:
```
RuntimeError: 
unknown type specifier:

        graph(%a : Float(123), %b : Float(4, 5, 6)):
          %c : NamedTuple(Tensor : Tuple, Tensor : Tuple) = prim::TupleConstruct(%a, %b)
               ~~~~~~~~~~ <--- HERE
          return (%c)
```

cc: @ke1337 @antoniojkim @wconstab @eellison

### CODE REVIEW

File: torch/csrc/jit/passes/refine_types.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace {
+static void VisitTupleNode(Node* node) {
+  TORCH_CHECK(
+      node->outputs().size() == 1, "Tuple must have exactly one output!");
+
+  Value* output = node->outputs()[0];
+  TupleType* tuple_type = dynamic_cast<TupleType*>(output->type().get());
+
+  TORCH_CHECK(tuple_type, "Existing result type must be a Tuple!");
```
Comment (eellison) [inline torch/csrc/jit/passes/refine_types.cpp:16]: nit: `output->type()->expectRef<TupleType>()` (will throw if it's not a tuple)

File: torch/csrc/jit/passes/refine_types.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  // Construct new tuple type based on input types.
+  if (tuple_type->names()) {
+    output->setType(c10::TupleType::createNamed(
+        tuple_type->name(), tuple_type->names().value(), types));
+  } else {
+    output->setType(c10::TupleType::create(types));
+  }
+}
```
Comment (eellison) [inline torch/csrc/jit/passes/refine_types.cpp:34]: I think `type->withContained(types)` should do this

**Review by eellison — COMMENTED** [review]
nice ! should we name this refine_tuple_types ? also, just curious, what is the use case from your end if anything in particular

**Comment by henrytwo** [comment]
> what is the use case from your end if anything in particular

In Torch MLIR have a `jit::Graph` that initially lacks shape information, and then we have a pass to insert it. What ended up happening though was that because the tuples were generated prior to the shapes, the types became out of sync

**Review by eellison — APPROVED** [review]
LGTM!

**Pull request closed without merging**